### PR TITLE
[CORDA-3098] - Close previous connection after reconnection

### PR DESCRIPTION
**Pull request by dimosr**
## Changes
This changes the `ReconnectingCordaRPCOps`/`ReconnectingObservable` to release resources on reconnection - it also enables back some tests that were flaky because of this.

Some context on some additional changes that are included:
- Adjusted `RpcProxyClientHandler` so that it also releases resources appropriately when connection is explicitly closed - this was already done on failover. See JIRA for more details around this
- Embedded a round-robin failover in `ReconnectingCordaRPCOps`. Until now, the round robin was done explicitly at a much lower level & using Artemis APIs. The issue with this is that there are a lot of different layers and it's rather hard to reason about. I realised that while troubleshooting the issues on this ticket, which was quite painful.
- Added a timeout on the latch, so that the test does not hang infinitely in failure situations.
- Switched assertions to use assertJ, so that when the test fails, more information is provided on what was the difference between the expected and the actual numbers.

## Testing
The `RpcReconnectTests` have now stopped being brittle - I've run them successfully 5 consecutive times on the lastest version - you can check TC history

### CODE REVIEW

**Comment by vkolomeyko** [comment]
Do we have an intention to merge this PR soon-ish?
Integration test `RPCStabilityTests.client doesnt leak threads when it fails to start` been failing for multiple days in OS `master` and CE `master` preventing integration into `master` branches.

**Comment by dimosr** [comment]
@vkolomeyko this PR is not ready to be merged. It resolves the issues with `RPCStabilityTests`, but unmasks other issues in `RpcReconnectTests`, which now become flaky (you can check TC history in this build to verify that). Further work needs to be done to identify the actual root cause and resolve that as well before merging.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt [hunk 1]
```diff
@@ -407,7 +407,24 @@ class RPCClientProxyHandler(
         }
 
         reaperScheduledFuture?.cancel(false)
```
Comment (JamesHR3) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: This class also has consumer/producer sessions - do these need to be cleaned up? (I didn't see anything in the code that does it, but I also only scanned it briefly so I may be missing something.)

Comment (JamesHR3) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Similarly with rpcProducer/Consumer.

Comment (dimosr) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: I think this is done by `sessionFactory.close()`/`sessionFactory.cleanup()` a few lines above.

Comment (JamesHR3) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Seems like the session factory should close the sessions when that's closed. I assume the rpcProducer/Consumer are closed when the session is closed? In which case this is probably fine.

**Pull request merged**